klippy/heater.py

```diff
@@ -141,6 +141,92 @@ def __init__(self, printer, config):
+
+    def _check_heating(self, eventtime):
+        next_time = 10.0 # next 10sec from now
+        with self.lock:
+            current_temp = self.last_temp
+            target_temp = self.target_temp
+        if self.protection_last_temp is None:
+            self.is_heating = False
+            self.is_runaway = False
+            self.is_cooling = False
+            # Set init value
+            self.protection_last_temp = current_temp
+            if (current_temp <= (target_temp - self.protect_hyst_runaway)):
+                self.is_heating = True
+                next_time = self.protection_period_heat
+            elif (target_temp < current_temp and 0 < target_temp):
+                self.is_cooling = True
+                next_time = self.protection_period
+            else:
+                self.is_runaway = True
+                next_time = self.protection_period
+        elif self.is_runaway:
+            # Check hysteresis during maintain
+            if (self.protect_hyst_runaway < abs(current_temp - target_temp)):
+                errorstr = "Thermal runaway! current temp {}, last {}". \
+                           format(current_temp, self.protection_last_temp)
+                self.set_temp(0, 0);
+                gcode = self.printer.lookup_object('gcode')
+                gcode.respond_error(errorstr)
+                self.printer.request_exit('shutdown')
+            self.protection_last_temp = current_temp
+            next_time = self.protection_period
+        elif (self.is_cooling):
+            next_time = self.protection_period_heat
+            if ((current_temp - self.protect_hyst_runaway) < target_temp):
+                self.is_cooling = False
+                self.is_heating = True
+        elif (self.is_heating):
+            # Check hysteresis during the preheating
+            if ((target_temp - self.protect_hyst_runaway) \
+                <= current_temp <=
+                (target_temp + self.protect_hyst_runaway)):
+                self.is_runaway = True;
+            elif (current_temp < target_temp):
+                if (abs(current_temp - self.protection_last_temp) < self.protection_hysteresis_heat):
+                    errorstr = "Heating error! current temp {}, last {}". \
+                               format(current_temp, self.protection_last_temp)
+                    self.set_temp(0, 0);
+                    gcode = self.printer.lookup_object('gcode')
+                    gcode.respond_error(errorstr)
+                    self.printer.request_exit('shutdown')
+            self.protection_last_temp = current_temp
+            next_time = self.protection_period_heat
+        logging.debug("check_heating(eventtime {}, next {}) {} / {}".
+                      format(eventtime, (eventtime + next_time),
+                             current_temp, target_temp))
+        return eventtime + next_time
+    def get_min_extrude_status(self):
```

Comment:
Oh, these get_min_extrude_status() and set_min_extrude_temp() need to be removed. 
I will fix this...